### ISSUE TITLE
Solver selecting worse rotations than Initial Guess

### ISSUE THREAD

**Issue by Envian** [dazemc/ffxiv-craft#29]
I am trying to use the solver to optimize a mid level craft. Every time I run it however, it immediately pulls the Trained Perfection + Groundwork combo to the beginning of the rotation, and produces a rotation with both less progress and quality than my initial guess.

I've attached my settings - which I hope contains the `[100] Cooking Mama` rotation. This rotation meets the quality and progress requirements to craft HQ, but running it through the solver will result in a rotation which meets neither.

Note that I've fiddled with my solver settings a lot over my time with this project - please let me know if something I changed caused this issue.

[settings(2).zip](https://github.com/user-attachments/files/16275199/settings.2.zip)

**Comment by dazemc** [comment]
TP just tries to prepend the first instance of groundwork or preparatory touch, I am working on a different method but I noticed other issues in doing so. I have noticed that the fitness and penalties that is supposed to keeping track of which gen has the best child is not reflecting. I also want to invalidate any sequences that are not valid, like ones that don't meet progress. There is "optimizations" that someone has written in the past that may have been beneficial at the time. Right now I'm working on a portion of the code that rewards sequences that are shorter because I think it might be the cause of some of these issues.

**Comment by dazemc** [comment]
Just pushed some changes, let me know if it looks better.

**Comment by Envian** [comment]
Thank you for the quick turnaround! Unfortunately I still get a similar problem even with the changes. Right now, the TP is being placed right after a veneration, pushing a groundwork out of the buff window. Here's a screenshot of the initial guess (Which hits 100% progress and quality), and the result:

![image](https://github.com/user-attachments/assets/34d11826-c53e-41cd-9b84-a6d9d6ec1576)

In this case, I expect the TP to be placed later - on a Groundwork which isn't covered by the Veneration buff. I can also imagine scenarios where TP isn't used, or is used on a 10 dura ability, simply because all the groundworks are covered by buffs which can't be delayed.

**Comment by dazemc** [comment]
> Thank you for the quick turnaround! Unfortunately I still get a similar problem even with the changes. Right now, the TP is being placed right after a veneration, pushing a groundwork out of the buff window. Here's a screenshot of the initial guess (Which hits 100% progress and quality), and the result:
> 
> ![image](https://private-user-images.githubusercontent.com/7389255/349763310-34d11826-c53e-41cd-9b84-a6d9d6ec1576.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEyNzI1NTQsIm5iZiI6MTcyMTI3MjI1NCwicGF0aCI6Ii83Mzg5MjU1LzM0OTc2MzMxMC0zNGQxMTgyNi1jNTNlLTQxY2QtOWI4NC1hNmQ5ZDZlYzE1NzYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcxOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MThUMDMxMDU0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzQ5MjlkMDNhMTI5OTlhOWY3NjU5Y2IzYjFiYzU4ZjU3MDIxYWY3MjRhODkwMDU1YTRjMzJhODY0MjJiNjM5ZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.2ZMUzPaMjYYFcPg0IwNWPruFpaDB1lU19LuzPRZKCrc)
> 
> In this case, I expect the TP to be placed later - on a Groundwork which isn't covered by the Veneration buff. I can also imagine scenarios where TP isn't used, or is used on a 10 dura ability, simply because all the groundworks are covered by buffs which can't be delayed.

Just pushed another update, I'm sick of seeing macros that do not complete progress!

**Comment by Envian** [comment]
Unfortunately I'm still stuck getting incomplete crafts. And the TP is still being placed incorrectly.

![image](https://github.com/user-attachments/assets/0094a175-d364-4365-8c7c-466a1552a323)

**Comment by Envian** [comment]
Hey dazemc,

I just now noticed your edit requesting the execution logs - Unfortunately I didn't save any of them from before, and the current solver isn't trying to incorporate TP at this time. 

One thing I did notice is that if i took the above two incomplete rotations, removed the instance of TP, the resulting rotation was both complete and fairly good. If I had to guess, fitness/completion was being calculated before TP was added to the rotation.

**Comment by dazemc** [comment]
Oh yeah I completely reworked how TP and the reward/penalty system works
for it. I forgot to ask if you had noticed a difference.

**Comment by Envian** [comment]
It looks like its almost fixed - In most scenarios TP is used properly, but only on longer crafts. Looks like its just not step efficent for short crafts, so it gets left out. I did come across one edge case:

![image](https://github.com/user-attachments/assets/70627215-e0ff-495a-bcb1-e59703012650)

In this solver, TP was added before veneration. Here's the execution log.

[TP Veneration.log](https://github.com/user-attachments/files/16420616/TP.Veneration.log)

Other than this edge case, it seems as if TP is being used properly. The result is correct - the craft will complete, it just wastes TP.
